x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -479,7 +535,60 @@ public void accessGranted(String requestId, Authentication authentication, Strin
+                    assert SetEnabledAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((SetEnabledRequest) msg).build();
+                } else if (msg instanceof ChangePasswordRequest) {
+                    assert ChangePasswordAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((ChangePasswordRequest) msg).build();
+                } else if (msg instanceof CreateApiKeyRequest) {
+                    assert CreateApiKeyAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((CreateApiKeyRequest) msg).build();
+                } else if (msg instanceof GrantApiKeyRequest) {
+                    assert GrantApiKeyAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((GrantApiKeyRequest) msg).build();
+                } else if (msg instanceof PutPrivilegesRequest) {
+                    assert PutPrivilegesAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((PutPrivilegesRequest) msg).build();
+                } else if (msg instanceof DeleteUserRequest) {
+                    assert DeleteUserAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((DeleteUserRequest) msg).build();
+                } else if (msg instanceof DeleteRoleRequest) {
+                    assert DeleteRoleAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((DeleteRoleRequest) msg).build();
+                } else if (msg instanceof DeleteRoleMappingRequest) {
+                    assert DeleteRoleMappingAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((DeleteRoleMappingRequest) msg).build();
+                } else if (msg instanceof InvalidateApiKeyRequest) {
+                    assert InvalidateApiKeyAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((InvalidateApiKeyRequest) msg).build();
+                } else if (msg instanceof DeletePrivilegesRequest) {
+                    assert DeletePrivilegesAction.NAME.equals(action);
+                    securityChangeLogEntryBuilder(requestId).withRequestBody((DeletePrivilegesRequest) msg).build();
+                } else {
+                    throw new IllegalStateException("Unknown message class type [" + msg.getClass().getSimpleName() +
+                            "] for the \"security change\" action [" + action + "]");
+                }
+            } catch (IOException e) {
+                // TODO ensure and test that IOExceptions are gracefully handled up the call stack
+                // TODO Especially check that all valid metadata can be serialized in the audit record
             }
```

Comment:
I have a general question: What is our intention for handling errors occur during auditing time? Do we want to propogate the error message back to user in the HTTP response, or do we only need to print it in server log? IIUC, by throwing IllegalStateException, the error will be in the HTTP response?

Reply:
> What is our intention for handling errors occur during auditing time?

I think we should handle auditing errors as any other business logic errors, i.e. return the error to the client in the HTTP response, and sometimes also log it.
We're not always rigorous about error handling for auditing. I'll try to get some tests for it. This is why I put these TODOs here for. 

> IIUC, by throwing IllegalStateException, the error will be in the HTTP response?

Yeah, because at some upper level in the call stack there will most likely be an `ActionListener.wrap` that forwards thrown exception when invoking `onSuccess` to the `onFailure` handler. This is a good defensive strategy, which worked well, but I'll try to add some explicit tests for the auditing case, because silent failures in this particular area are technically bugs (unlike the other places where's it's a nuisance to the debugging person on the SDH).

Reply:
Thanks for the explanation Albert. If the intention is to report the error back to users, it makes sense to throw exception. My question was mainly about whether it is necessary to alert user with a logging error. Since audit logging is an opt-in feature, I can be convinced that faliure should be more visible. But I wonder how clients should handle this error when it occurs. I assume this will give back a status code `5xx`? Since it is not `4xx`, clients may decide to retry the request. The API key related requests are not idempotent which could be a bit anonying. Other requests are idempotent, but sometimes have the side effects of clearing relevant caches.